File: 300Plan/_020MergeSort.java
public class _020MergeSort {
    private void sort(int[] arr, int left, int right) {
        if (left < right) {
            int mid = left + (right - left ) / 2;

            // divide 成两部分
            sort(arr, left, mid);
            sort(arr, mid + 1, right);

            // 合并两部分
            merge(arr, left, right, mid);
        }
    }
    private void merge(int[] arr, int left, int right, int mid) {
        int lLen = mid - left + 1;
        int rLen = right - mid;
        int[] leftPart = new int[lLen];
        int[] rightPart = new int[rLen];

        for (int i = 0; i < lLen; i++) {
            leftPart[i] = arr[left + i];
        }
        for (int i = 0; i < rLen; i++) {
            rightPart[i] = arr[mid + 1 + i];
        }

        int lIndex = 0, rIndex = 0, index = left;

        while (lIndex < lLen && rIndex < rLen) {
            if (leftPart[lIndex] <= rightPart[rIndex]) {
                arr[index] = leftPart[lIndex];
                lIndex++;
            } else {
                arr[index] = rightPart[rIndex];
                rIndex++;
            }
            index++;
        }
        // 检查还未排序的
        while (lIndex < lLen) {
            arr[index] = leftPart[lIndex];
            lIndex++;
            index++;
        }
        while (rIndex < rLen) {
            arr[index] = rightPart[rIndex];
            rIndex++;
            index++;
        }
    }

    public static void main(String[] args) {
        _020MergeSort test = new _020MergeSort();
        int arr[] = { 12, 11, 13, 5, 6, 7 };
        for (int num : arr) {
            System.out.print(num + " ");
        }
        System.out.println();
        test.sort(arr, 0, arr.length - 1);
        for (int num : arr) {
            System.out.print(num + " ");
        }
    }
}

File: 300Plan/_020MergeSort.py
def merge_sort(lst):
    lst_sorted = lst
    if len(lst_sorted) > 1:
        mid = len(lst_sorted) // 2
        leftPart = lst_sorted[:mid]
        rightPart = lst_sorted[mid:]

        # 先递归直至只有两个数
        merge_sort(leftPart)
        merge_sort(rightPart)

        lIndex = rIndex = index = 0
        while lIndex < len(leftPart) and rIndex < len(rightPart):
            if leftPart[lIndex] < rightPart[rIndex]:
                lst_sorted[index] = leftPart[lIndex]
                lIndex += 1
            else:
                lst_sorted[index] = rightPart[rIndex]
                rIndex += 1
            index += 1

        while lIndex < len(leftPart):
            lst_sorted[index] = leftPart[lIndex]
            lIndex += 1
            index += 1
        while rIndex < len(rightPart):
            lst_sorted[index] = rightPart[rIndex]
            rIndex += 1
            index += 1
    return lst_sorted


lst = [12, -1, 100, 5, 2, 17]
print(merge_sort(lst))
